File: CS/XtraGridTutorial.Step4/MainForm.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.Linq;
using System.Text;
using System.Windows.Forms;
using DataLayer;
using System.Diagnostics;
using DevExpress.XtraEditors;
using DataLayer.Data.Classes;
using DevExpress.Utils;
using DevExpress.XtraEditors.Repository;

namespace XtraGridTutorial {
    public partial class MainForm : XtraForm {
        private UnitOfWork uow;
        public MainForm()
        {
            InitializeComponent();
            uow = new UnitOfWork();
            SetDataSource();
            InitControls();
        }

        private void barButtonSave_ItemClick(object sender, DevExpress.XtraBars.ItemClickEventArgs e)
        {
            uow.Commit();
        }

        private void barButtonReload_ItemClick(object sender, DevExpress.XtraBars.ItemClickEventArgs e)
        {
            uow.ReloadContext();
            SetDataSource();
        }

        private void SetDataSource()
        {
            accountBindingSource.DataSource = uow.GetData<Account>();
            categoryBindingSource.DataSource = uow.GetData<Category>();
            transactionBindingSource.DataSource = uow.GetData<Transaction>();
        }
        private void InitControls()
        {
            InitImageComboBox(repositoryItemImageComboBox1);
            InitImageComboBox(repositoryItemImageComboBox2);
        }

        private void InitImageComboBox(RepositoryItemImageComboBox item)
        {
            item.Items.AddEnum(typeof(TransactionType));
            for (int i = 0; i < item.Items.Count; i++)
            {
                item.Items[i].ImageIndex = i;
            }
        }

        private void transactionsGridView_ValidateRow(object sender, DevExpress.XtraGrid.Views.Base.ValidateRowEventArgs e)
        {
            Transaction row = (Transaction)e.Row;
            transactionsGridView.ClearColumnErrors();
            if (row.TrasactionDate != null && row.TrasactionDate > DateTime.Today)
            {
                e.Valid = false;
                transactionsGridView.SetColumnError(colTrasactionDate, ValidationErrorStringsHelper.GetDateErrorMessage(colTrasactionDate.GetCaption()));
            }
            if (row.Account == null)
            {
                e.Valid = false;
                transactionsGridView.SetColumnError(colAccount, ValidationErrorStringsHelper.GetAccountErrorMessage(colAccount.GetCaption()));
            }
        }

        private void transactionsGridView_InvalidRowException(object sender, DevExpress.XtraGrid.Views.Base.InvalidRowExceptionEventArgs e)
        {
            e.ExceptionMode = DevExpress.XtraEditors.Controls.ExceptionMode.NoAction;
        }
    }
}

File: VB/XtraGridTutorial.Step7/MainForm.vb
Imports Microsoft.VisualBasic
Imports System
Imports System.Collections.Generic
Imports System.ComponentModel
Imports System.Data
Imports System.Drawing
Imports System.Linq
Imports System.Text
Imports System.Windows.Forms
Imports DataLayer
Imports System.Diagnostics
Imports DevExpress.XtraEditors
Imports DataLayer.Data.Classes
Imports DevExpress.Utils
Imports DevExpress.XtraEditors.Repository

Namespace XtraGridTutorial
	Partial Public Class MainForm
		Inherits XtraForm
		Private uow As UnitOfWork
		Public Sub New()
			InitializeComponent()
			uow = New UnitOfWork()
			SetDataSource()
			InitControls()
		End Sub

		Private Sub barButtonSave_ItemClick(ByVal sender As Object, ByVal e As DevExpress.XtraBars.ItemClickEventArgs) Handles barButtonSave.ItemClick
			uow.Commit()
		End Sub

		Private Sub barButtonReload_ItemClick(ByVal sender As Object, ByVal e As DevExpress.XtraBars.ItemClickEventArgs) Handles barButtonReload.ItemClick
			uow.ReloadContext()
			SetDataSource()
		End Sub

		Private Sub SetDataSource()
			accountBindingSource.DataSource = uow.GetData(Of Account)()
			categoryBindingSource.DataSource = uow.GetData(Of Category)()
			transactionBindingSource.DataSource = uow.GetData(Of Transaction)()
		End Sub
		Private Sub InitControls()
			InitImageComboBox(repositoryItemImageComboBox1)
			InitImageComboBox(repositoryItemImageComboBox2)
		End Sub

		Private Sub InitImageComboBox(ByVal item As RepositoryItemImageComboBox)
			item.Items.AddEnum(GetType(TransactionType))
			For i As Integer = 0 To item.Items.Count - 1
				item.Items(i).ImageIndex = i
			Next i
		End Sub

		Private Sub transactionsGridView_ValidateRow(ByVal sender As Object, ByVal e As DevExpress.XtraGrid.Views.Base.ValidateRowEventArgs) Handles transactionsGridView.ValidateRow
			Dim row As Transaction = CType(e.Row, Transaction)
			transactionsGridView.ClearColumnErrors()
            If row.TrasactionDate > DateTime.Today Then
                e.Valid = False
                transactionsGridView.SetColumnError(colTrasactionDate, ValidationErrorStringsHelper.GetDateErrorMessage(colTrasactionDate.GetCaption()))
            End If
			If row.Account Is Nothing Then
				e.Valid = False
				transactionsGridView.SetColumnError(colAccount, ValidationErrorStringsHelper.GetAccountErrorMessage(colAccount.GetCaption()))
			End If
		End Sub

		Private Sub transactionsGridView_InvalidRowException(ByVal sender As Object, ByVal e As DevExpress.XtraGrid.Views.Base.InvalidRowExceptionEventArgs) Handles transactionsGridView.InvalidRowException
			e.ExceptionMode = DevExpress.XtraEditors.Controls.ExceptionMode.NoAction
		End Sub

		Private Sub barButtonItem1_ItemClick(ByVal sender As Object, ByVal e As DevExpress.XtraBars.ItemClickEventArgs) Handles barButtonItem1.ItemClick
			transactionsGridControl.ShowPrintPreview()
		End Sub
	End Class
End Namespace
